cmake: Do not modify `CMAKE_TRY_COMPILE_TARGET_TYPE` globally

hebasto opened the pull request:
This was requested in https://github.com/bitcoin/bitcoin/pull/31359#issuecomment-2515287092.

From https://github.com/bitcoin/bitcoin/pull/31359#issuecomment-2511246212:
> (Almost?) every CMake check internally uses the [`try_compile()`](https://cmake.org/cmake/help/latest/command/try_compile.html) command, whose behaviour, in turn, depends on the [`CMAKE_TRY_COMPILE_TARGET_TYPE`](https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html) variable:
> 
>    1. The default value, `EXECUTABLE`, enables both compiler and linker checks.
> 
>    2. The `STATIC_LIBRARY` value enables only compiler checks.
> 
> 
> To mimic Autotools' behaviour, we [disabled](https://github.com/bitcoin/bitcoin/blob/d3f42fa08fc385752881afa5740f40287cfb4d8b/cmake/module/CheckSourceCompilesAndLinks.cmake#L9-L10) linker checks by setting `CMAKE_TRY_COMPILE_TARGET_TYPE` to `STATIC_LIBRARY` globally (perhaps not the best design). This effectively separates the entire CMake script into regions where `CMAKE_TRY_COMPILE_TARGET_TYPE` is:
> 
>    * unset
> 
>    * set to `STATIC_LIBRARY`
> 
>    * set to `EXECUTABLE`

From https://github.com/bitcoin/bitcoin/pull/31359#issuecomment-2515287092:
> > This seems very fragile and unintuitive, and the fact that this could silently break at any point is not documented in any way. I don't think other bad design decisions should lead to us having to write even more boilerplate code to fix things that should "just work" (minus the upstream bugs).
> 
> Agreed. I forgot that we set `CMAKE_TRY_COMPILE_TARGET_TYPE` globally. And even worse, it's buried in a module. If that upsets CMake internal tests, I think we should undo that.

This PR ensures that `CMAKE_TRY_COMPILE_TARGET_TYPE` is modified only within local scopes.

Additionally, the `FUZZ_LIBS` variable has been introduced to handle additional libraries required for linking, rather than link options, in certain build environment, such as OSS-Fuzz.

theuni commented:
High-level question: Do we actually need the old autotools behavior? If this is just about shaving off some time when running the linker, I'd rather skip that optimization. Is there actually a case where a static-lib compile-test would succeed, an executable test would fail, and _also_ we would be ok with that failure?

If not, we could just skip the complexity and not modify the policy at all.

hebasto commented:
> High-level question: Do we actually need the old autotools behavior?

That was a strict requirement for the staging branch, and mapping theAutotools' `AC_PREPROC_IFELSE`, `AC_COMPILE_IFELSE` and `AC_LINK_IFELSE` macros to CMake functions was essential.

> Is there actually a case where a static-lib compile-test would succeed, an executable test would fail, and _also_ we would be ok with that failure?

I don't think so.

> Is there an obvious reason I'm missing to not just use CMake's default behavior here rather than fighting it?

I'd love to live with CMake's default behavior!

> It was .4 sec slower, but that's not enough to care about :)

It might be even slower with LTO enabled, though.

hebasto commented:
> Proposed alternative to this PR: https://github.com/theuni/bitcoin/commits/cmake-trycompile-exe/
> What do you think?

Thank you. It Looks great!

I also pushed an updated branch, which has a cleaner resulting code.

However, I'm happy to switch to your branch, if you prefer it.

hebasto commented:
A typo has been fixed (

theuni commented on CMakeLists.txt:
```diff
@@ -372,12 +372,13 @@ target_link_options(sanitize_interface INTERFACE ${SANITIZER_LDFLAGS})
       extern \"C\" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { return 0; }
       // No main() function.
     " FUZZ_BINARY_LINKS_WITHOUT_MAIN_FUNCTION
+    CXXFLAGS ${SANITIZER_LDFLAGS}
```
Could you add an LDFLAGS param to the function as `LINK_OPTIONS` and use that here instead?

hebasto replied:
`LDFLAGS` might make more sense because, for example, in the OSS-Fuzz environment, the `SANITIZER_LDFLAGS` variable can be set to `-fsanitize=fuzzer` or `/usr/lib/libFuzzingEngine.a`.

hebasto replied:
The `LDFLAGS` option has been [added](https://github.com/bitcoin/bitcoin/pull/31662#issuecomment-2593891189).

theuni replied:
> OSS-Fuzz

Wait, they're using a completely undocumented var. That shouldn't be any reason for us to change our behavior here. That's waaaay to brittle for us to have to worry about, no?

Why are they not using `CMAKE_EXE_LINKER_FLAGS` or `LDFLAGS` ?

hebasto replied:
> Why are they not using `CMAKE_EXE_LINKER_FLAGS` or `LDFLAGS` ?

Both variants will work though.

cc the OSS-Fuzz crowd: @maflcko @dergoegge @fanquake @achow101

theuni replied:
Is it not an issue to link the fuzzer lib with our non-fuzz binaries? If not, yeah, let's just do that :)

hebasto replied:
> Is it not an issue to link the fuzzer lib with our non-fuzz binaries?

I don't think so: https://github.com/bitcoin/bitcoin/blob/2d07384243c9552aa8e95c80d7a279e2d224a753/CMakeLists.txt#L211-L231

theuni replied:
Oh, right! Still.. I could imagine it breaking CMake checks. But sure, if it works, that'd be better than hacking around it here.

theuni commented on cmake/module/TestAppendRequiredLibraries.cmake:
```diff
@@ -25,10 +25,11 @@ function(test_append_socket_library target)
+  check_cxx_source_compiles("${check_socket_source}" IFADDR_LINKS_WITHOUT_LIBSOCKET)
   if(NOT IFADDR_LINKS_WITHOUT_LIBSOCKET)
-    check_cxx_source_links_with_libs(socket "${check_socket_source}" IFADDR_NEEDS_LINK_TO_LIBSOCKET)
+    set(CMAKE_REQUIRED_LIBRARIES socket)
```
Could you add a `LINK_LIBRARIES` option to `check_cxx_source_compiles_with_flags` and use that here instead?

hebasto replied:
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/31662#issuecomment-2593891189).

hebasto commented:
@theuni

Thank you for your review! Your comments have been addressed.

theuni commented on cmake/module/CheckSourceCompilesAndLinks.cmake:
```diff
@@ -12,13 +12,33 @@ macro(check_cxx_source_links source)
+function(check_cxx_source_compiles_with_flags source result_var)
+  cmake_parse_arguments(PARSE_ARGV 2 _ "" "" "CXXFLAGS;LDFLAGS;LINK_LIBRARIES")
+  list(JOIN __CXXFLAGS " " CMAKE_REQUIRED_FLAGS)
+  set(CMAKE_REQUIRED_LIBRARIES ${__LDFLAGS} ${__LINK_LIBRARIES})
```
Why does `CMAKE_REQUIRED_LIBRARIES` get `LDFLAGS` rather than setting `LINK_OPTIONS`?

hebasto replied:
> I think a static lib would work fine as an option though?

Correct. This is documented by CMake.

theuni replied:
I meant as `LINK_OPTIONS`. I'd rather have LDFLAGS forward to that unless there's a reason not to.

hebasto replied:
Okay, I've checked CMake docs again.

`CMAKE_REQUIRED_LINK_OPTIONS` is forwarded to `target_link_options()`, which is [designed](https://cmake.org/cmake/help/latest/command/target_link_options.html) "to add any link options".

I'll update the PR shortly.

hebasto replied:
> I meant as `LINK_OPTIONS`. I'd rather have LDFLAGS forward to that unless there's a reason not to.

In that case a static library (e.g., `/usr/lib/libFuzzingEngine.a`) would be placed _before_ object files being linked, and the linker might fail.

theuni replied:
Yeah, I really don't think we should be worrying about their busted flags.

For OSS-Fuzz, we can change it to pass:
`-DSANITIZER_LDFLAGS="-Wl,$LIB_FUZZING_ENGINE"`. That should fix the ordering problem if it's an absolute file while still working with "-lFuzzingEngine" as well.

theuni replied:
Blah, ok, sorry. I've been messing around with this. Final answer:

- Let's add a `SANITIZER_LIBS` var
- add BOTH `LINK_LIBRARIES` and `LDFLAGS` to check_cxx_source_compiles_with_flags
- pass them both for the sanitizers
- update oss-fuzz to use `SANITIZER_LIBS` instead.

That way our stuff is well-defined and makes sense and we're not coding around oss-fuzz, but we still expose the option we need there.

hebasto replied:
cc the OSS-Fuzz crowd: @maflcko @dergoegge @fanquake @achow101

theuni replied:
@fanquake Are you ok if we break OSS-Fuzz here with a proper fix (to avoid future technical debt), and PR the one-liner to fix it there post-merge?

fanquake replied:
Yes

fanquake commented:
https://github.com/bitcoin/bitcoin/pull/31662/checks?check_run_id=37171033299:
```bash
[10:53:16.049] gmake[2]: Entering directory '/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu'
[10:53:16.093] [ 72%] Linking CXX executable fuzz
[10:53:16.093] cd /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src/test/fuzz && /usr/bin/cmake -E cmake_link_script CMakeFiles/fuzz.dir/link.txt --verbose=1
[10:53:16.117] /usr/bin/clang++-20 -ftrivial-auto-var-init=pattern -O2 -g -fsanitize=fuzzer,address,undefined,float-divide-by-zero,integer -fstack-protector-all -fcf-protection=full -fstack-clash-protection -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -fPIE -pie CMakeFiles/fuzz.dir/addition_overflow.cpp.o CMakeFiles/fuzz.dir/addrman.cpp.o CMakeFiles/fuzz.dir/asmap.cpp.o CMakeFiles/fuzz.dir/asmap_direct.cpp.o CMakeFiles/fuzz.dir/autofile.cpp.o CMakeFiles/fuzz.dir/banman.cpp.o CMakeFiles/fuzz.dir/base_encode_decode.cpp.o CMakeFiles/fuzz.dir/bech32.cpp.o CMakeFiles/fuzz.dir/bip324.cpp.o CMakeFiles/fuzz.dir/bitdeque.cpp.o CMakeFiles/fuzz.dir/bitset.cpp.o CMakeFiles/fuzz.dir/block.cpp.o CMakeFiles/fuzz.dir/block_header.cpp.o CMakeFiles/fuzz.dir/block_index.cpp.o CMakeFiles/fuzz.dir/blockfilter.cpp.o CMakeFiles/fuzz.dir/bloom_filter.cpp.o CMakeFiles/fuzz.dir/buffered_file.cpp.o CMakeFiles/fuzz.dir/chain.cpp.o CMakeFiles/fuzz.dir/checkqueue.cpp.o CMakeFiles/fuzz.dir/cluster_linearize.cpp.o CMakeFiles/fuzz.dir/coins_view.cpp.o CMakeFiles/fuzz.dir/coinscache_sim.cpp.o CMakeFiles/fuzz.dir/connman.cpp.o CMakeFiles/fuzz.dir/crypto.cpp.o CMakeFiles/fuzz.dir/crypto_aes256.cpp.o CMakeFiles/fuzz.dir/crypto_aes256cbc.cpp.o CMakeFiles/fuzz.dir/crypto_chacha20.cpp.o CMakeFiles/fuzz.dir/crypto_chacha20poly1305.cpp.o CMakeFiles/fuzz.dir/crypto_common.cpp.o CMakeFiles/fuzz.dir/crypto_diff_fuzz_chacha20.cpp.o CMakeFiles/fuzz.dir/crypto_hkdf_hmac_sha256_l32.cpp.o CMakeFiles/fuzz.dir/crypto_poly1305.cpp.o CMakeFiles/fuzz.dir/cuckoocache.cpp.o CMakeFiles/fuzz.dir/decode_tx.cpp.o CMakeFiles/fuzz.dir/descriptor_parse.cpp.o CMakeFiles/fuzz.dir/deserialize.cpp.o CMakeFiles/fuzz.dir/eval_script.cpp.o CMakeFiles/fuzz.dir/feefrac.cpp.o CMakeFiles/fuzz.dir/fee_rate.cpp.o CMakeFiles/fuzz.dir/feeratediagram.cpp.o CMakeFiles/fuzz.dir/fees.cpp.o CMakeFiles/fuzz.dir/flatfile.cpp.o CMakeFiles/fuzz.dir/float.cpp.o CMakeFiles/fuzz.dir/golomb_rice.cpp.o CMakeFiles/fuzz.dir/headerssync.cpp.o CMakeFiles/fuzz.dir/hex.cpp.o CMakeFiles/fuzz.dir/http_request.cpp.o CMakeFiles/fuzz.dir/i2p.cpp.o CMakeFiles/fuzz.dir/integer.cpp.o CMakeFiles/fuzz.dir/key.cpp.o CMakeFiles/fuzz.dir/key_io.cpp.o CMakeFiles/fuzz.dir/kitchen_sink.cpp.o CMakeFiles/fuzz.dir/load_external_block_file.cpp.o CMakeFiles/fuzz.dir/locale.cpp.o CMakeFiles/fuzz.dir/merkleblock.cpp.o CMakeFiles/fuzz.dir/message.cpp.o CMakeFiles/fuzz.dir/miniscript.cpp.o CMakeFiles/fuzz.dir/minisketch.cpp.o CMakeFiles/fuzz.dir/mini_miner.cpp.o CMakeFiles/fuzz.dir/muhash.cpp.o CMakeFiles/fuzz.dir/multiplication_overflow.cpp.o CMakeFiles/fuzz.dir/net.cpp.o CMakeFiles/fuzz.dir/net_permissions.cpp.o CMakeFiles/fuzz.dir/netaddress.cpp.o CMakeFiles/fuzz.dir/netbase_dns_lookup.cpp.o CMakeFiles/fuzz.dir/node_eviction.cpp.o CMakeFiles/fuzz.dir/overflow.cpp.o CMakeFiles/fuzz.dir/p2p_handshake.cpp.o CMakeFiles/fuzz.dir/p2p_headers_presync.cpp.o CMakeFiles/fuzz.dir/p2p_transport_serialization.cpp.o CMakeFiles/fuzz.dir/package_eval.cpp.o CMakeFiles/fuzz.dir/parse_hd_keypath.cpp.o CMakeFiles/fuzz.dir/parse_iso8601.cpp.o CMakeFiles/fuzz.dir/parse_numbers.cpp.o CMakeFiles/fuzz.dir/parse_script.cpp.o CMakeFiles/fuzz.dir/parse_univalue.cpp.o CMakeFiles/fuzz.dir/partially_downloaded_block.cpp.o CMakeFiles/fuzz.dir/policy_estimator.cpp.o CMakeFiles/fuzz.dir/policy_estimator_io.cpp.o CMakeFiles/fuzz.dir/poolresource.cpp.o CMakeFiles/fuzz.dir/pow.cpp.o CMakeFiles/fuzz.dir/prevector.cpp.o CMakeFiles/fuzz.dir/primitives_transaction.cpp.o CMakeFiles/fuzz.dir/process_message.cpp.o CMakeFiles/fuzz.dir/process_messages.cpp.o CMakeFiles/fuzz.dir/protocol.cpp.o CMakeFiles/fuzz.dir/psbt.cpp.o CMakeFiles/fuzz.dir/random.cpp.o CMakeFiles/fuzz.dir/rbf.cpp.o CMakeFiles/fuzz.dir/rolling_bloom_filter.cpp.o CMakeFiles/fuzz.dir/rpc.cpp.o CMakeFiles/fuzz.dir/script.cpp.o CMakeFiles/fuzz.dir/script_assets_test_minimizer.cpp.o CMakeFiles/fuzz.dir/script_descriptor_cache.cpp.o CMakeFiles/fuzz.dir/script_flags.cpp.o CMakeFiles/fuzz.dir/script_format.cpp.o CMakeFiles/fuzz.dir/script_interpreter.cpp.o CMakeFiles/fuzz.dir/script_ops.cpp.o CMakeFiles/fuzz.dir/script_parsing.cpp.o CMakeFiles/fuzz.dir/script_sigcache.cpp.o CMakeFiles/fuzz.dir/script_sign.cpp.o CMakeFiles/fuzz.dir/scriptnum_ops.cpp.o CMakeFiles/fuzz.dir/secp256k1_ec_seckey_import_export_der.cpp.o CMakeFiles/fuzz.dir/secp256k1_ecdsa_signature_parse_der_lax.cpp.o CMakeFiles/fuzz.dir/signature_checker.cpp.o CMakeFiles/fuzz.dir/signet.cpp.o CMakeFiles/fuzz.dir/socks5.cpp.o CMakeFiles/fuzz.dir/span.cpp.o CMakeFiles/fuzz.dir/string.cpp.o CMakeFiles/fuzz.dir/strprintf.cpp.o CMakeFiles/fuzz.dir/system.cpp.o CMakeFiles/fuzz.dir/timeoffsets.cpp.o CMakeFiles/fuzz.dir/torcontrol.cpp.o CMakeFiles/fuzz.dir/transaction.cpp.o CMakeFiles/fuzz.dir/txdownloadman.cpp.o CMakeFiles/fuzz.dir/tx_in.cpp.o CMakeFiles/fuzz.dir/tx_out.cpp.o CMakeFiles/fuzz.dir/tx_pool.cpp.o CMakeFiles/fuzz.dir/txorphan.cpp.o CMakeFiles/fuzz.dir/txrequest.cpp.o CMakeFiles/fuzz.dir/utxo_snapshot.cpp.o CMakeFiles/fuzz.dir/utxo_total_supply.cpp.o CMakeFiles/fuzz.dir/validation_load_mempool.cpp.o CMakeFiles/fuzz.dir/vecdeque.cpp.o CMakeFiles/fuzz.dir/versionbits.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/coincontrol.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/coinselection.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/crypter.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/fees.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/notifications.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/scriptpubkeyman.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/spend.cpp.o CMakeFiles/fuzz.dir/__/__/wallet/test/fuzz/wallet_bdb_parser.cpp.o -o fuzz  util/libtest_fuzz.a ../../libbitcoin_cli.a ../../libbitcoin_common.a ../../util/libbitcoin_util.a ../../../libminisketch.a ../../../libleveldb.a ../../univalue/libunivalue.a ../../secp256k1/lib/libsecp256k1.a ../../wallet/libbitcoin_wallet.a ../util/libtest_util.a ../../libbitcoin_node.a ../../../libminisketch.a ../../../libleveldb.a ../../../libcrc32c.a ../../../libcrc32c_sse42.a /usr/lib/x86_64-linux-gnu/libevent_extra.so /usr/lib/x86_64-linux-gnu/libevent_core.so /usr/lib/x86_64-linux-gnu/libevent_pthreads.so /usr/lib/x86_64-linux-gnu/libevent.so ../../libbitcoin_common.a ../../util/libbitcoin_util.a ../../libbitcoin_consensus.a ../../secp256k1/lib/libsecp256k1.a ../../crypto/libbitcoin_crypto.a ../../crypto/libbitcoin_crypto_sse41.a ../../crypto/libbitcoin_crypto_avx2.a ../../crypto/libbitcoin_crypto_x86_shani.a ../../univalue/libunivalue.a /usr/lib/x86_64-linux-gnu/libsqlite3.so  
[10:53:16.586] /usr/bin/ld: util/libtest_fuzz.a(fuzz.cpp.o): in function `main':
[10:53:16.600] /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src/test/fuzz/util/./test/fuzz/fuzz.cpp:237: multiple definition of `main'; /usr/lib/llvm-20/lib/clang/20/lib/linux/libclang_rt.fuzzer-x86_64.a(FuzzerMain.cpp.o):(.text.main+0x0): first defined here
[10:53:28.557] clang++-20: error: linker command failed with exit code 1 (use -v to see invocation)
[10:53:28.564] gmake[2]: *** [src/test/fuzz/CMakeFiles/fuzz.dir/build.make:2238: src/test/fuzz/fuzz] Error 1
[10:53:28.564] gmake[2]: Leaving directory '/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu'
[10:53:28.565] gmake[1]: *** [CMakeFiles/Makefile2:998: src/test/fuzz/CMakeFiles/fuzz.dir/all] Error 2
[10:53:28.565] gmake[1]: Leaving directory '/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu'
[10:53:28.566] gmake: *** [Makefile:136: all] Error 2
```

theuni commented on CMakeLists.txt:
```diff
@@ -383,17 +383,17 @@ if(SANITIZERS)
       extern \"C\" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) { return 0; }
       // No main() function.
     " FUZZ_BINARY_LINKS_WITHOUT_MAIN_FUNCTION
-    LDFLAGS ${SANITIZER_LDFLAGS}
```
This still needs `LDFLAGS ${SANITIZER_LDFLAGS}` as before in order for the check to pass, so I think the `sanitizer_interface` changes above should be undone.

hebasto replied:
Fixed.

theuni reviewed (approved):
utACK 9e7eb06a24b670e8e72e0233a63bbc67733d4adb

A reminder that this will require a [change in OSS-Fuzz](https://github.com/google/oss-fuzz/blob/master/projects/bitcoin-core/build.sh#L74):
`s/SANITIZER_LDFLAGS/FUZZ_LIBS/`

TheCharlatan commented on CMakeLists.txt:
```diff
@@ -383,9 +384,9 @@ if(SANITIZERS)
     message(FATAL_ERROR "Linker did not accept requested flags, you are missing required libraries.")
   endif()
 endif()
-target_link_options(sanitize_interface INTERFACE ${SANITIZER_LDFLAGS})
```
In commit 9e7eb06a24b670e8e72e0233a63bbc67733d4adb
Nit: Is this change related to the actual change introducing `FUZZ_LIBS` in this commit?

hebasto replied:
Unrelated changes have been dropped.

sedited reviewed (approved):
ACK 9e7eb06a24b670e8e72e0233a63bbc67733d4adb

Can you add the note about `FUZZ_LIBS` usage to the PR description?

dergoegge commented:
>  s/SANITIZER_LDFLAGS/FUZZ_LIBS/

Happy to take care of this once this gets merged.

theuni commented on cmake/module/CheckSourceCompilesAndLinks.cmake:
```diff
@@ -6,9 +6,6 @@ include_guard(GLOBAL)
-set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
-
 macro(check_cxx_source_links source)
   set(CMAKE_TRY_COMPILE_TARGET_TYPE EXECUTABLE)
```
Probably not here. IIRC hebasto split this out of a larger commit. Agree that it's confusing, but since it's resolved later I'm not too worried. I'd say it's up to hebasto if he wants to fix it for the sake of history or not.

hebasto replied:
Thanks! Reworked.

hebasto commented:
> Can you add the note about `FUZZ_LIBS` usage to the PR description?

Sure! Added.

hebasto committed:
build: Don't override CMake's default try_compile target

CMake assumes the default value internally, so overriding this causes
problems. The minimal speedup of skipping the linker isn't worth the
complexity of setting it to static.

hebasto committed:
cmake: Delete `check_cxx_source_links_with_flags` macro

hebasto committed:
cmake: Convert `check_cxx_source_compiles_with_flags` to a function

hebasto committed:
cmake: Delete `check_cxx_source_links*` macros

hebasto committed:
scripted-diff: Rename CMake helper module

-BEGIN VERIFY SCRIPT-
git mv cmake/module/CheckSourceCompilesAndLinks.cmake cmake/module/CheckSourceCompilesWithFlags.cmake
sed -i 's|\<CheckSourceCompilesAndLinks\>|CheckSourceCompilesWithFlags|g' $(git grep -l 'CheckSourceCompilesAndLinks')
-END VERIFY SCRIPT-

hebasto committed:
cmake: Introduce `FUZZ_LIBS`

CMake distinguishes recommended methods for handling (1) linker options
and (2) libraries used during linking. Therefore, it is both reasonable
and consistent to introduce a dedicated variable for the latter,
particularly when a build environment, such as OSS-Fuzz, requires
linking against additional libraries.

hebasto commented:
Addressed the latest @TheCharlatan's feedback.

sedited reviewed (approved):
Re-ACK 2c4b229c906de6250500d3af2b44808e90b9ce0b

theuni reviewed (approved):
utACK 2c4b229c906de6250500d3af2b44808e90b9ce0b

I didn't re-review each commit in great detail, but the diff from before is as I'd expect.

fanquake commented:
Opened https://github.com/google/oss-fuzz/pull/13081.
